src/app/api/v1/customers/sign-up/route.ts

```diff
@@ -0,0 +1,120 @@
+import { NextResponse } from 'next/server'
+import to from 'await-to-js'
+import { AuthError, PostgrestError } from '@supabase/supabase-js'
+import { ZodError } from 'zod'
+
+import { createClient } from '@/utils/supabase/server'
+import { customerSignupFormSchema } from '@/types/forms'
+import { stripe } from '@/server/services/stripe'
+
+type Response =
+  | {
+      success: boolean
+      message?: string
+    }
+  | {
+      success: false
+      message: string
+      error: Error
+    }
+
+export const POST = async (req: Request): Promise<NextResponse<Response>> => {
+  const { error } = await handleCustomerSignup(req)
+
+  if (error instanceof ZodError) {
+    return NextResponse.json(
+      { success: false, message: error.message, error },
+      { status: 400 },
+    )
+  }
+
+  if (error instanceof AuthError) {
+    return NextResponse.json(
+      {
+        success: false,
+        message: error.message,
+        error,
+      },
+      { status: error.status },
+    )
+  }
+
+  if (error) {
+    return NextResponse.json(
+      { success: false, message: error.message, error },
+      { status: 500 },
+    )
+  }
+
+  return NextResponse.json({ success: true })
+}
+
+async function handleCustomerSignup(
+  req: Request,
+): Promise<
+  | { success: false; error: Error | PostgrestError }
+  | { success: true; error: null }
+> {
+  const supabase = createClient()
+
+  const validationResult = customerSignupFormSchema.safeParse(
+    await req.formData(),
+  )
+
+  if (validationResult.success === false) {
+    return { success: false, error: validationResult.error }
+  }
+
+  const {
+    email,
+    password,
+    preferred_name: preferredName,
+    fullname,
+  } = validationResult.data
+
+  const [error, data] = await to(
+    stripe.customers.create({
+      email,
+      name: fullname,
+    }),
+  )
+
+  if (error) {
+    return { success: false, error }
+  }
+
+  const stripeCustomer = data
+
+  const signupResult = await supabase.auth.signUp({
+    email,
+    password,
+    options: {
+      data: {
+        role: 'customer',
+        customer_id: stripeCustomer.id,
+        onboarded: true,
+        preferred_name: preferredName,
+        fullname,
+      },
+    },
+  })
+
+  if (signupResult.error) {
+    return { success: false, error: signupResult.error }
+  }
+
+  const userId = signupResult.data?.user?.id as string
+
+  const customerInsertResult = await supabase
+    .from('customers')
+    .insert([{ user_id: userId, preferred_name: preferredName, fullname }])
```

Comment:
Would love for this to be wrapped as a SQL transaction with the `auth user create` so we can process this as an all-or-nothing operation.

Unfortunately supabase does not support transactions outright. `auth.users` are also treated differently, so even if we could, there would be roadblocks along the way.

Related Information: https://stackoverflow.com/a/77052403

Reply:
There's a chance that the user will be created in supabase, but the Customer record would not be created. We should think of a way to fix this.

Let's come back to this later, since I this is a low-frequency edge case at the moment. But I'm thinking of adding a `verify-onboarding` endpoint that will fix this. _Might be the way to go, not sure._